platform/ios/PolyPodApp/OIDAuth/OIDAuth.swift

```diff
@@ -0,0 +1,137 @@
+    }
+    
+    /// Ends user session by clearing the auth state
+    func clearAuthState() {
+        self.state = nil
+        _ = OIDTokenStorage
+            .removeAuthState(forService: authorizationRequest.clientID)
+            .inspectError { err in
+                Log.error("Failed to remove auth state \(err)")
+            }
+        // TBD end session request
+    }
+    
+    /// Loads the Auth state.
+    /// - Parameter completion: The result of the load operation.
+    ///                         After a successful load, the `state` peroperty will be initialized.
+    func loadAuthState(completion: @escaping (Result<Void, Error>) -> Void) {
+        if let state = try? OIDTokenStorage.getAuthState(forService: authorizationRequest.clientID).get() {
+            self.state = state
+            completion(.success(()))
+            return
+        }
+        
+        self.currentAuthFlow = OIDAuthState.authState(
+            byPresenting: authorizationRequest,
+            externalUserAgent: oidExternalUserAgent) { state, error in
+            completion(Result {
+                switch (state, error) {
+                case let (state?, nil):
+                    self.state = state
+                case let (nil, error?):
+                    throw error
+                default:
+                    throw InvalidAuthResponseRepresentation()
+                }
+            })
+        }
+    }
+    
+    /// Handles Auth redirect to complete the Auth flow.
+    /// After redirect was handled, any subsequent calls to this function will be ignored,
+    /// unless a new auth flow is started.
+    /// - Parameter url: The auth redirect URL
+    func handleAuthRedirect(_ url: URL) {
+        currentAuthFlow?.resumeExternalUserAgentFlow(with: url)
+        currentAuthFlow = nil
+    }
+}
+
+extension OIDAuth: OIDAuthStateChangeDelegate {
+    // Handle state changes when token is refreshed or invalidated.
+    func didChange(_ state: OIDAuthState) {
+        self.state = state
+    }
+}
+
+extension OIDAuth {
+    /// An user agent which does not trigger any actual authorization flow.
+    /// Meant to be used when the authorization flow happens on the backend side,
+    /// while the app "just" needs to handle the code/token exchange.
+    /// This is necessary to "trick" AppAuth that a full OAuth2.0 flow is happening form the client side.
+    private final class OIDExternalUserAgentNoOp: NSObject, OIDExternalUserAgent {
+        func present(_ request: OIDExternalUserAgentRequest, session: OIDExternalUserAgentSession) -> Bool {
+            return true
+        }
+        
+        func dismiss(animated: Bool, completion: @escaping () -> Void) {
+            completion()
+        }
+    }
+    
+    /// Membership feature Auth configuration
+    static func membershipFeatureAuth() -> OIDAuth {
+        let redirectURL = URL(string: "coop.polypoly.polypod://membership_feature/oauth")!
```

Comment:
is this protocol://domain?

Reply:
yes, where the scheme is the app bundle id.

Reply:
Can't you read the app bundle scheme from somewhere?
Point is: literals...

Reply:
Yes, that would be an improvement.